Fixes Issue #14

Deathwing777 opened the pull request:
Fixes: "Needs a global table so other mods can be nether aware. #14"

Deathwing777 committed:
Fixes Issue #14

Fixes: "Needs a global table so other mods can be nether aware. #14"

Deathwing777 committed:
Fixes HybridDog#16

Added new chat function `in_hell [<player_name>]` and used it to find the bug. The issue was that `local function obsidian_teleport(player, pname, target)` failed to test `nether.trap_players` before modifying `players_in_nether`.

Deathwing777 committed:
Code cleanup and new global/chat function.

- Added a new global function to the nether table that lets external mods tell us to update a player's Nether status based on the player's position. This function has been exposed to the console via a registered chat function.
- Switched from NETHER as the global table to nether as the global table after discovering that nether was already being used in init.lua.
- Switched from accessing function calls in the global table from  table syntax to object syntax; as it looks cleaner.

HybridDog commented on nether/portal.lua:
```diff
@@ -29,6 +29,27 @@ do
+	else
+		players_in_nether[pname] = nil
+	end
+end
```
Style:
I think the `save_nether_players` function, which saves the `nether_players` file should directly follow the `do … end` code which loads the `nether_players` file since these two code pieces are related. I suggest to put the `nether['is_player_in_nether']` and `nether['external_nether_teleport']` function definitions after the `save_nether_players`.

Deathwing777 replied:
Okay.

HybridDog commented on nether/portal.lua:
```diff
@@ -29,6 +29,27 @@ do
 
+-- Nether aware mods will need to know if a player is in the nether.
+nether['is_player_in_nether'] = function (player)
+	local pname = player:get_player_name()
+	if players_in_nether[pname] == nil then
+		return false
+	else
+		return true
+	end
```
Style:
The whole function can be simplified to 
```lua
return players_in_nether[player:get_player_name()]
```
or, if it is necessary that the function returns false instead of nil, to 
```lua
return players_in_nether[player:get_player_name()] or false
```

Deathwing777 replied:
The longer form is more readable at a glance. I generally try to avoid cramming too much into one line, but this is your mod, so of course I will adhere to your style guidelines.

HybridDog commented on nether/portal.lua:
```diff
@@ -29,6 +29,27 @@ do
 end
 
+-- Nether aware mods will need to know if a player is in the nether.
+nether['is_player_in_nether'] = function (player)
```
Style: `nether['is_player_in_nether'] = function (player)` can be simplified to 
`function nether.is_player_in_nether(player)`.

Deathwing777 replied:
same as nether['external_nether_teleport']. I have no problem changing it.

HybridDog commented on nether/portal.lua:
```diff
@@ -29,6 +29,27 @@ do
+end
+
+-- Nether aware mods may have other means of moving players between the Nether and Overworld, and if so, they should tell us about it so we can keep track of the player state.
+nether['external_nether_teleport'] = function (player)
```
Style: this can be simplified to `function nether.external_nether_teleport(player)`

Deathwing777 replied:
Yeah, I already changed all of the usages of this function to the object style references. I left the declaration as a table style reference because I thought it important to help others just learning Lua to understand that there are multiple ways to use the table object, and the table style declaration was relatively clean. I have no problem changing it if you prefer it that way.

HybridDog commented on nether/portal.lua:
```diff
@@ -29,6 +29,27 @@ do
+	end
+end
+
+-- Nether aware mods may have other means of moving players between the Nether and Overworld, and if so, they should tell us about it so we can keep track of the player state.
```
Style: The line is too long. Please break the comment into multiple lines. My personal preference is at most 80 characters per line.

Deathwing777 replied:
No problem.

HybridDog commented on nether/portal.lua:
```diff
@@ -29,6 +29,27 @@ do
+nether['external_nether_teleport'] = function (player)
+	local pos = player:get_pos()
+	local pname = player:get_player_name()
+	if (pos.y < nether.start) and (pos.y >= nether.bottom) then
```
Everywhere else in the code, a player is in `players_in_nether` if he/she is below `nether.start` independent of the `nether.bottom` value. For consistency, I suggest to use `if pos.y < nether.start then` here, too.

HybridDog commented on nether/portal.lua:
```diff
@@ -79,11 +100,13 @@ end
-	players_in_nether[pname] = true
-	save_nether_players()
-	update_background(player, true)
+	if nether.trap_players then
```
If I understand my messy code correctly, this is the wrong way to fix #16. 
Before the changes, if damage is enabled, the obsidian portal kills players even if `nether.trap_players` is disabled. If `nether.trap_players` is disabled, the `minetest.register_on_respawnplayer` is not executed, so the died player remains in the overworld.

So instead of changing `obsidian_teleport`, I'd change `obsi_teleport_player` so that it executes `obsidian_teleport` without target position only if both damage and `nether.trap_players` are enabled and otherwise executes it with target position:
```
	local has_teleported
	if damage_enabled and nether.trap_players then
		obsidian_teleport(player, pname)
		has_teleported = true
	elseif not mclike_portal then
		local target = vector.round(get_player_died_target(player))
		if generated_or_generate(target) then
			obsidian_teleport(player, pname, target)
			has_teleported = true
		end
	end
```
I have not tested if my suggestion works.

Deathwing777 replied:
Okay. Found it. I tried to stay out of the messy stuff near the bottom. :D

If I understand this correctly, solving #16 this way will no longer kill the player if they try to use an Obsidian portal with `nether.trap_players = false`. Is that what you intended? It makes sense to me, but my previous impression is that you wanted to kill any player attempting to use an Obsidian portal, no matter what (assuming damage enabled).

My next commit will have your suggested solution. Let me know if that is not what you intended, and I will switch it back.

HybridDog replied:
Yes, this is what I intended, thanks. 
If damage is enabled and `nether.trap_players` is disabled and the Obsidian portal kills the player, the player cannot use it to get to the nether, so I think in this case the player should be teleported instead of killed.

HybridDog commented on nether/portal.lua:
```diff
@@ -176,6 +199,62 @@ minetest.register_chatcommand("from_hell", {
+			return false, "Something went wrong."
+		end
+		
+		status = pname.." is in the "
```
>     portal.lua:219:3: setting non-standard global variable status

`status` should be changed to a local variable.

Deathwing777 replied:
oops. Still getting used to Lua.

HybridDog commented on nether/portal.lua:
```diff
@@ -176,6 +199,62 @@ minetest.register_chatcommand("from_hell", {
+		end
+		
+		nether.external_nether_teleport(player)
+		status = pname.." is in the "
```
>     portal.lua:247:3: setting non-standard global variable status

Deathwing777 replied:
oops. Still getting used to Lua.

HybridDog commented on nether/portal.lua:
```diff
@@ -176,6 +199,62 @@ minetest.register_chatcommand("from_hell", {
+		else
+			status = status.."OVERWORLD!"
+		end
+		
```
Style:
Here are two tabs in the empty line.
This is not the only line with trailing whitespace, as reported by luacheck:

>    portal.lua:218:1: line contains only whitespace
    portal.lua:225:1: line contains only whitespace
    portal.lua:245:1: line contains only whitespace
    portal.lua:253:1: line contains only whitespace

Deathwing777 replied:
Python habits.... Odd that lua would complain about whitespace only lines rather than just silently stripping them out.

HybridDog replied:
Lua doesn't complain but [luacheck](https://github.com/lunarmodules/luacheck/) does, which is a static analyzer and linter for Lua code.

HybridDog commented on nether/portal.lua:
```diff
@@ -176,6 +199,62 @@ minetest.register_chatcommand("from_hell", {
+	params = "[<player_name>]",
+	description = "Update player state if they got to or from the nether in another way.",
+	func = function(name, pname)
+		if not minetest.check_player_privs(name, {nether=true}) then
+			return false,
+				"You need the nether priv to execute this chatcommand."
+		end
```
I think a cleaner/simpler way to check privileges is using the `privs` field in the chat command definition. The chat commands which already exist before this Pull Request also don't use `privs` field for some reason.

Deathwing777 replied:
I copied and pasted the chat command above it as a template. This is exactly the same way that `from_hell` and `to_hell` check privs. I am not sure what you mean by `privs` field. Is this something that you are wanting me to fix prior to accepting a pull request?

HybridDog replied:
Here's an example where the `privs` field is used to check if the player has the `worldedit` privilege: https://github.com/red-001/more_chatcommands/blob/4faf1c5e5a710ad907cebc76c7ccd1419f31ae4d/init.lua#L178

Since the other chat commands also use `minetest.check_player_privs` instead of the `privs` field, for this Pull Request both ways to check the privilege are fine for me.

HybridDog reviewed (changes requested):
I don't think the `nether.external_nether_teleport` function is useful for other mods. If `nether.trap_players` is enabled, the nether mod blocks player teleportation functions so that they cannot teleport a player to/from the nether.

Deathwing777 commented:
> I don't think the `nether.external_nether_teleport` function is useful for other mods. If `nether.trap_players` is enabled, the nether mod blocks player teleportation functions so that they cannot teleport a player to/from the nether.

If nether.trap_players == true, then yes it is not useful. In the game world I am making for my kids, I have this set to false as being trapped is no fun for the kiddos. In that world, I am running your nether mod along side Shadow Realms, and Meseportals. Currently with the upstream code, using a Meseportal to teleport to the nether will cause the player to start receiving 1 HP/sec damage caused by the Shadow Realms mod. To fix this, Meseportals needs to be able to tell the nether mod that we have teleported to below the nether generation ceiling. This external function seemed the simplest way of making the mods play nice together.

Deathwing777 committed:
Implemented HybridDog's change requests.

This should now be ready to merge.

Deathwing777 commented:
Now the real question is how to handle someone getting in or out of the nether the hard way (e.g. digging).  :D

Even with my contributions, if `nether.trap_players = true` and a player digs their way out of the nether, they will still be listed in `players_in_nether`, and the next time they try to use a Meseportal they will go right back to the nether. 

Obviously you don't want code running all the time checking `y pos` if it can be avoided. The chat command `update_hells_registry` that I added can work as a stopgap, but I only added that to help with debugging/testing, so it is intended to eventually go away once things are stable. I will have a closer look at the code you use to block teleporation. I suspect that has something to do with the emergent phenomenon.

HybridDog commented:
> Now the real question is how to handle someone getting in or out of the nether the hard way (e.g. digging). :D

I think the same problem happens when a player uses /sethome in the nether and teleports back there after leaving the nether.

If `nether.trap_players` is disabled, the player can freely teleport to or from the nether, so in this case the player position determines if the player is in the nether or not.
On the other hand, if `nether.trap_players` is enabled, the player should only be in the nether if and only if he/she is in `players_in_nether`.
So, perhaps `nether.is_player_in_nether` can be changed to this: 
```lua
function nether.is_player_in_nether(player)
	if nether.trap_players then
		return players_in_nether[player:get_player_name()]
	end
	local pos = player:get_pos()
	return pos.y < nether.start
end
```

If `nether.trap_players` is enabled, external mods cannot (at least not easily) teleport a player to or from the nether, so perhaps `nether.external_nether_teleport` can be changed to a function which teleports the player and sets them into or out of the nether:
```lua
function nether.external_nether_teleport(player, pos)
	if not nether.trap_players then
		player:set_pos(pos)
		return
	end
	local destination_in_nether = pos.y < nether.start
	update_background(player, destination_in_nether)
	local pname = player:get_player_name()
	players_in_nether[pname] = destination_in_nether or nil
	player:set_pos(pos)
end
```

Deathwing777 commented:
> > Now the real question is how to handle someone getting in or out of the nether the hard way (e.g. digging). :D
> 
> I think the same problem happens when a player uses /sethome in the nether and teleports back there after leaving the nether.
> 
> If `nether.trap_players` is disabled, the player can freely teleport to or from the nether, so in this case the player position determines if the player is in the nether or not. On the other hand, if `nether.trap_players` is enabled, the player should only be in the nether if and only if he/she is in `players_in_nether`. So, perhaps `nether.is_player_in_nether` can be changed to this:
> 
> ```lua
> function nether.is_player_in_nether(player)
> 	if nether.trap_players then
> 		return players_in_nether[player:get_player_name()]
> 	end
> 	local pos = player:get_pos()
> 	return pos.y < nether.start
> end
> ```
> 
> If `nether.trap_players` is enabled, external mods cannot (at least not easily) teleport a player to or from the nether, so perhaps `nether.external_nether_teleport` can be changed to a function which teleports the player and sets them into or out of the nether:
> 
> ```lua
> function nether.external_nether_teleport(player, pos)
> 	if not nether.trap_players then
> 		player:set_pos(pos)
> 		return
> 	end
> 	local destination_in_nether = pos.y < nether.start
> 	update_background(player, destination_in_nether)
> 	local pname = player:get_player_name()
> 	players_in_nether[pname] = destination_in_nether or nil
> 	player:set_pos(pos)
> end
> ```

Missing tests against the bottom of the nether, but otherwise looks good. Since player names should get stored if and only if they are trapped in the nether, I suggest the `players_in_nether` table be renamed to `players_trapped_in_nether` as that makes it much clearer.

Deathwing777 committed:
Implemented HybridDog's change request.

- Modifications to `nether.is_player_in_nether()` and `nether.external_nether_teleport()`
- Renamed `players_in_nether[]` to `players_trapped_in_nether[]`
- Fixed punctuation in a comment. =D

Deathwing777 commented:
While I was looking at your code that blocks teleporation, I noticed that there are some local functions that only get defined if `nether.trap_players == true`. It is probably a good idea to define those functions regardless. I can think of a few interesting use cases where `nether.trap_players` would need to change on the fly rather than just be a configuration option requiring a restart of the game world. Examples:

1) The long standing configuration of astronomical bodies creates a strong veil between the Overworld and the Nether, but during certain short duration astronomical events the veil is weakened. During times of thin veil, players can escape the nether, but so can the hell-beasts (so nether mobs start spawning in the overworld near obsidian portals).

2) Obsidian portals work just like they do in Minecraft, but somewhere near the world spawn there is a nether fortress with loot chests, etc. In the heart of the treasure room, there is a pedestal with a unique powerful item (weapon, armor, etc). If this item is removed from the pedestal the room gets filled with a purple aura and clues appear on the walls about how to construct a non-obsidian portal ( and `nether.trap_players = true`). Returning the item to the pedestal fixes the portals (`nether.trap_players = false`). If a player dies with the item in their inventory, or if they attempt to place the item into a chest, then the item disappears and returns to it's pedestal in the nether.

3) Players get trapped in the nether, and must construct the new style portal to get out. Once a player escapes with the new style portal, the portal is destroyed; but that particular player can then use obsidian portals to and from the nether.

4) Everything was beautifully peaceful in the Overworld, and adventurers routinely explore the nether to find excitement. Then one day, everything changes (Curse over the land | Someone defiled a temple and angered the gods | Someone removed a power object from a pedestal | etc). Now obsidian portals start spawning everywhere in the Overworld, and hordes of demons start wreaking havoc across the land. With every passing day the numbers and strength of the demons invading the Overworld increase. The king of the land places a bounty on the demons, but there is little hope of turning the tide of destruction. How many days do we have left before Balrogs wander the land in mass? We need a group of brave adventurers to delve into the Nether and find a way to put this to a stop. Unfortunately, they discover that the only way to achieve this is to xxxxxx which also breaks the return functionality of obsidian portals (`nether.trap_players = true`). We saved the world, but now we need to figure out how to save ourselves...

Technically 3) doesn't need changes to `nether.trap_players`. That would best be implemented with a `priv`. I just wanted to throw that idea out there anyway.

Deathwing777 commented:
> While I was looking at your code that blocks teleporation, I noticed that there are some local functions that only get defined if `nether.trap_players == true`. It is probably a good idea to define those functions regardless.

Unless I am missing something, implementing that is as simple as removing the `if nether.trap_players then` from:

`-- Disallow teleportation and change spawn positions if the nether traps players
if nether.trap_players then
        -- randomly set player position when he/she dies in nether
        minetest.register_on_respawnplayer(function(player)
                ...
        end)

        -- override set_pos etc, to disallow player teleportion by e.g. travelnet
        local function can_teleport(player, pos)`
                ....
        end

        minetest.register_on_joinplayer(function(player)
                ....
        end
end`

Deathwing777 commented:
I will test today's commit this evening when I get home.

Deathwing777 commented:
Upon testing, it looks like your change to `nether.is_player_in_nether(player)` does not work as intended. This function literally just needs to check if the player is between the floor and ceiling of nether generation. Adding the test for the player name in the table if `nether.trap_players == true` breaks things as digging through the nether no longer gets detected and portals are now the only way that the function changes it's return value. One block below or above nether generation needs to `return false`, anywhere within the nether generation region needs to `return true`.

Deathwing777 commented:
Testing also uncovered either another issue, or a gap in my understanding of your code in `init.lua`.

```
-- Height of the nether (bottom of the nether is nether_middle - NETHER_HEIGHT)
local NETHER_HEIGHT = 30
```

I had used the information in that comment to add `nether.bottom = nether_middle - NETHER_HEIGHT`. Apparently I forgot to commit that change to github, so it is currently only on my local machine. I will fix that with my next commit. Testing shows that nether generation goes lower than `y=-20030`. The bottom of the nether on this map is at `y=-20190`. What am I missing?

Also, I think it would be a good idea to move a few of these settings to `settings.lua` to expose them to the Minetest settings GUI. Specifically, I am thinking about `nether_middle` and `NETHER_HEIGHT`. What do you think?

Deathwing777 committed:
Made changes after testing previous commit.

init.lua:
- Added `nether.bottom`

portal.lua:
- Reverted `nether.is_player_in_nether(player)` to previous state.
- Added `nether.is_player_trapped_in_nether(player)`
- Added 2 new chat functions for debugging purposes.
- Tested that chat function `/update_hells_registry` indeed works to check if a player has dug their way out of the nether.
- Moved the functional code from `/update_hells_registry` to `nether.registry_update(player)`
- Fixed some English grammar in automated chat messages.

Deathwing777 commented:
Done with testing for this evening. The current set of changes seems to work well (other than the issue with `nether.bottom` not actually being the bottom of the nether. I moved the functional bits of code from `/update_hells_registry` to `nether.update_registry(player)` as it has been tested and verified to correctly detect when a player has escaped the nether by digging. This could be called with a `minetest.register_globalstep()` function, but that might not be the optimal way to do it.

Deathwing777 committed:
Fixed broken elseif

- Lua is weird.

HybridDog commented on nether/portal.lua:
```diff
@@ -41,6 +41,47 @@ local function save_nether_players()
+	local in_nether = (pos.y < nether.start) and (pos.y >= nether.bottom)
+	local pname = player:get_player_name()
+	if nether.trap_players then
+		players_trapped_in_nether[pname] = in_nether
```
`save_nether_players` saves all keys in `players_trapped_in_nether` to the nether_players file even if the value is false, so `players_trapped_in_nether[pname]` should either be unset (nil) or true.
```lua
		players_trapped_in_nether[pname] = in_nether or nil
```

Alternatively, the `save_nether_players` function can be changed to check if the value evaluates to a truth value before saving the player name.

Deathwing777 replied:
Fixed

HybridDog commented on nether/portal.lua:
```diff
@@ -41,6 +41,47 @@ local function save_nether_players()
+		return
+	end
+	local destination_in_nether = (pos.y < nether.start) and (pos.y >= nether.bottom)
+	update_background(player, destination_in_nether)
```
`update_background` is out of scope here since the `local update_background` is further down in the code.

>    portal.lua:64:2: accessing undefined variable update_background
>    portal.lua:78:3: accessing undefined variable update_background
>    portal.lua:81:3: accessing undefined variable update_background

Deathwing777 replied:
Okay. I moved that up above the functions that call `update_background()`

HybridDog commented on nether/init.lua:
```diff
@@ -52,6 +52,9 @@ nether.start = f_h_max+100
 local NETHER_HEIGHT = 30
 
+-- bottom height of the nether.
+nether.bottom = nether_middle - NETHER_HEIGHT
```
>  Testing shows that nether generation goes lower than y=-20030. The bottom of the nether on this map is at y=-20190. What am I missing?

The nether mapgen code currently does not use a clean way to determine the bottom of the nether, so it depends on the mapchunk size, which is usually 80 nodes and can be changed with the [`chunksize` setting](https://github.com/minetest/minetest/blob/fbec378869691444e354425073a7210555a9b31d/minetest.conf.example#L3047-L3054).

With 80 nodes mapchunk size, the bottom could be between `NETHER_BOTTOM-179` and `NETHER_BOTTOM-116` or something like that, so I think `nether_middle - NETHER_HEIGHT - 100` could fit a bit closer to the actual bottom.

Deathwing777 replied:
Thanks.

HybridDog commented:
> While I was looking at your code that blocks teleporation, I noticed that there are some local functions that only get defined if nether.trap_players == true. It is probably a good idea to define those functions regardless. I can think of a few interesting use cases where nether.trap_players would need to change on the fly rather than just be a configuration option requiring a restart of the game world.

Trapping players in the nether by overriding the teleport functions is already quite hacky in my opinion. If dynamically overriding `nether.trap_players` was possible, the overriding can no longer be completely disabled by a setting and it may be difficult to implement it without accidentally breaking compatibility with other mods. 

> I think it would be a good idea to move a few of these settings to settings.lua to expose them to the Minetest settings GUI. Specifically, I am thinking about nether_middle and NETHER_HEIGHT. What do you think?

I agree, but this should be done in separate Pull Requests.

Deathwing777 commented:
> > While I was looking at your code that blocks teleporation, I noticed that there are some local functions that only get defined if nether.trap_players == true. It is probably a good idea to define those functions regardless. I can think of a few interesting use cases where nether.trap_players would need to change on the fly rather than just be a configuration option requiring a restart of the game world.
> 
> Trapping players in the nether by overriding the teleport functions is already quite hacky in my opinion. If dynamically overriding `nether.trap_players` was possible, the overriding can no longer be completely disabled by a setting and it may be difficult to implement it without accidentally breaking compatibility with other mods.
> 
> > I think it would be a good idea to move a few of these settings to settings.lua to expose them to the Minetest settings GUI. Specifically, I am thinking about nether_middle and NETHER_HEIGHT. What do you think?
> 
> I agree, but this should be done in separate Pull Requests.

Okay. If everything looks good to you, sounds like a good time to merge. After the merge I will have a crack at moving some of the settings from `init.lua` to `settings.lua`.

Deathwing777 committed:
Implemented HybridDog's change requests.

- Should be ready to merge now.

HybridDog commented:
Yes, I've squashed and merged it. Thanks.
